tron/api/async_resource.py

```diff
@@ -0,0 +1,48 @@
+import threading
+
+from twisted.internet import threads
+from twisted.web import server
+
+
+class AsyncResource():
+    capacity = 10
+    semaphore = threading.Semaphore(value=capacity)
+    lock = threading.Lock()
+
+    @staticmethod
+    def finish(result, request):
+        request.write(result)
+        request.finish()
+
+    @staticmethod
+    def process(fn, resource, request):
+        with AsyncResource.semaphore:
+            return fn(resource, request)
+
+    @staticmethod
+    def bounded(fn):
+        def wrapper(resource, request):
+            d = threads.deferToThread(
+                AsyncResource.process, fn, resource, request
+            )
+            d.addCallback(AsyncResource.finish, request)
+            d.addErrback(lambda f: f)
+            return server.NOT_DONE_YET
+
+        return wrapper
+
+    @staticmethod
+    def exclusive(fn):
+        def wrapper(*args, **kwargs):
+            # ensures only one exclusive request starts consuming the semaphore
+            with AsyncResource.lock:
+                # this will wait until all bounded requests finished processing
+                for _ in range(AsyncResource.capacity):
```

Comment:
This might reduce the concurrency of the system. For example, it would acquire 9 semaphores and wait for the last one. During that period, the concurrency decreases to 1. And the period for acquiring all semaphores would become longer if we increase the capacity. What if we don't wait until all bounded requests finished? Maybe it's too aggressive...

Reply:
my goal here was to ensure that during any POST request no GET requests are being processed because that could cause data race issues. we can relax this constraint later as we implement more granular locks around critical pieces of data.